### PR TITLE
feat(pruner): shared deletion limit

### PR DESCRIPTION
**Pull request by shekhirin**
With the addition of headers and transactions pruning for the needs of snapshots, it becomes even harder to manage the batch sizes for pruner, so we both not hit the MDBX freelist issues and still prune the data.

This PR refactors the batch sizes into one shared limit for maximum number of deleted entries per pruner run. Each prune part deletes as much as possible, it decreases the deletion limit, and further parts use new limit. It also means that prune parts now have different priority, because e.g. if receipts, which are pruned first, consume the deletion limit fully, consequent parts won't prune anything.

Tested on both mainnet and sepolia, example log
```console
2023-10-04T18:12:00.521326Z  INFO reth::node::events: Pruner finished tip_block_number=4424545 elapsed=899.944121ms parts={SenderRecovery: (Finished, 0), Receipts: (Finished, 0), ContractLogs: (Finished, 2588), AccountHistory: (Finished, 2976), StorageHistory: (Finished, 18652)}
```

### CODE REVIEW

File: crates/consensus/beacon/src/engine/test_utils.rs [hunk 1]
```diff
@@ -520,7 +520,7 @@ where
             self.base_config.chain_spec.clone(),
             5,
             PruneModes::none(),
-            PruneBatchSizes::default(),
+            MAINNET.prune_delete_limit,
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/test_utils.rs:523]: can this be used from the `chain_spec`?

File: crates/prune/src/pruner.rs [hunk 2]
```diff
@@ -31,6 +31,13 @@ use tracing::{debug, error, instrument, trace};
 /// and `false` if there's more data to prune in further runs.
 pub type PrunerResult = Result<bool, PrunerError>;
 
+/// Result of part pruning.
+///
+/// Returns `true` if pruning has been completed up to the target block,
+/// and `false` if there's more data to prune in further runs. Also returns number of entries pruned
+/// (deleted from the database).
+type PrunePartResult = Result<(bool, usize), PrunerError>;
```
Comment (mattsse) [inline crates/prune/src/pruner.rs:39]: I'm not a fan of these primitive tuple return values, especially for bools, 
can we replace the bool with an enum?

**Review by mattsse — CHANGES_REQUESTED** [review]
the explanation makes sense to me.

style nits.

haven't reviewed the provider changes in detail yet, ptal @joshieDo

File: crates/consensus/beacon/src/engine/test_utils.rs (hunk 1 above)
Comment (shekhirin) [reply crates/consensus/beacon/src/engine/test_utils.rs]: absolutely, fixed

File: crates/prune/src/pruner.rs (hunk 2 above)
Comment (shekhirin) [reply crates/prune/src/pruner.rs]: agree, introduced a enum with two variants

**Review by mattsse — APPROVED** [review]
lgtm

**Comment by joshieDo** [comment]
PR itself lgtm. `last_pruned_block_number` becomes a confusing name if we consider snapshots though (since it will be at a higher height than the first block to prune from the snapshot) , and we might want to change it in the future.

File: crates/prune/src/pruner.rs [hunk 3]
```diff
@@ -95,11 +95,16 @@ impl<DB: Database> Pruner<DB> {
...
+        let mut parts = BTreeMap::new();
 
         // TODO(alexey): prune snapshot parts of data (headers, transactions)
         // let highest_snapshots = *self.highest_snapshots_tracker.borrow();
 
+        let mut delete_limit = self.delete_limit *
+            self.last_pruned_block_number
+                .map_or(1, |last_pruned_block_number| tip_block_number - last_pruned_block_number)
+                as usize;
+
```
Comment (joshieDo) [inline crates/prune/src/pruner.rs:107]: can we just add a doc emphasizing that `last_pruned_block_number` is close to `block_interval` and cannot be too high in the current state?

**Review by joshieDo — APPROVED** [review]

**Pull request merged**